service-rs/httpgate/src/proxy.rs

```diff
@@ -0,0 +1,522 @@
+pub struct DevboxProxy {
+    registry: Arc<DevboxRegistry>,
+    /// Host parser for extracting protocol, uniqueID and port from Host header
+    host_parser: HostParser,
+}
+
+impl DevboxProxy {
+    pub const fn new(registry: Arc<DevboxRegistry>, agent_port: u16) -> Self {
+        Self {
+            registry,
+            host_parser: HostParser::new(agent_port),
+        }
+    }
+
+    /// Resolve the backend address from uniqueID.
+    ///
+    /// Performs a two-step lookup:
+    /// 1. uniqueID -> DevboxInfo (namespace, devbox_name)
+    /// 2. namespace/devbox_name -> pod_ip
+    ///
+    /// Returns:
+    /// - `BackendResult::Ok` if uniqueID is registered and Pod IP is available
+    /// - `BackendResult::NotFound` if uniqueID is not registered
+    /// - `BackendResult::NotRunning` if uniqueID is registered but Pod IP is not available
+    fn resolve_backend(&self, unique_id: &str, port: u16) -> BackendResult {
+        // Step 1: Look up devbox info
+        let Some(info) = self.registry.get_devbox(unique_id) else {
+            return BackendResult::NotFound;
+        };
+
+        // Step 2: Look up pod IP
+        let Some(pod_ip) = self.registry.get_pod_ip(&info.namespace, &info.devbox_name) else {
+            return BackendResult::NotRunning;
+        };
+
+        debug!(
+            unique_id = %unique_id,
+            namespace = %info.namespace,
+            devbox_name = %info.devbox_name,
+            pod_ip = %pod_ip,
+            port = port,
+            "Resolved backend"
+        );
+
+        BackendResult::Ok(pod_ip, port)
+    }
+
+    /// Send a 404 Not Found response
+    async fn send_not_found(session: &mut Session) -> Result<bool> {
+        let mut header = ResponseHeader::build(404, None)?;
+        header.insert_header("Content-Length", BODY_NOT_FOUND.len().to_string())?;
+        header.insert_header("Content-Type", "text/plain")?;
+        session
+            .write_response_header(Box::new(header), false)
+            .await?;
+        session
+            .write_response_body(Some(BODY_NOT_FOUND.into()), true)
+            .await?;
+        Ok(true)
+    }
+
+    /// Send a 503 Service Unavailable response (devbox not running)
+    async fn send_service_unavailable(session: &mut Session) -> Result<bool> {
+        let mut header = ResponseHeader::build(503, None)?;
+        header.insert_header("Content-Length", BODY_NOT_RUNNING.len().to_string())?;
+        header.insert_header("Content-Type", "text/plain")?;
+        session
+            .write_response_header(Box::new(header), false)
+            .await?;
+        session
+            .write_response_body(Some(BODY_NOT_RUNNING.into()), true)
+            .await?;
+        Ok(true)
+    }
+}
+
+#[async_trait]
+impl ProxyHttp for DevboxProxy {
+    type CTX = Option<ProxyCtx>;
+
+    fn new_ctx(&self) -> Self::CTX {
+        None
+    }
+
+    async fn request_filter(&self, session: &mut Session, ctx: &mut Self::CTX) -> Result<bool> {
```

Comment:
Nit: If somehow the implementation of `resolve_backend` changes and it takes a long time, `request_filter` might not yield to the executor.